File: kartwheel/app/src/main/java/us/handstand/kartwheel/KartWheel.java
package us.handstand.kartwheel;


import android.app.Application;
import android.database.Cursor;

import com.squareup.sqldelight.SqlDelightStatement;

import java.io.IOException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import us.handstand.kartwheel.model.AndroidStorage;
import us.handstand.kartwheel.model.Database;
import us.handstand.kartwheel.model.User;
import us.handstand.kartwheel.network.API;

public class KartWheel extends Application implements Interceptor {
    private static final ExecutorService es = Executors.newCachedThreadPool();
    private static User user;
    private final OkHttpClient okHttpClient = new OkHttpClient.Builder().addInterceptor(this).build();

    @Override
    public void onCreate() {
        super.onCreate();
        AndroidStorage.initialize(this);
        Database.initialize(this);
        API.initialize(okHttpClient, "http://10.0.0.173:3000");

        submitWork(() -> {
            SqlDelightStatement getUserStatement = User.FACTORY.select_all(AndroidStorage.getString(AndroidStorage.USER_ID));
            try (Cursor cursor = Database.get().query(getUserStatement.statement, getUserStatement.args)) {
                if (cursor.moveToFirst()) {
                    user = User.SELECT_ALL_MAPPER.map(cursor);
                }
            }
        });
    }

    public static User getUser() {
        return user;
    }

    public static Future submitWork(Runnable runnable) {
        return es.submit(runnable);
    }

    @Override
    public Response intercept(Chain chain) throws IOException {
        Request.Builder requestBuilder = chain.request().newBuilder();
        requestBuilder.header("Content-Type", "application/json");
        requestBuilder.header("Accept", "application/json");
        return chain.proceed(requestBuilder.build());
    }
}

File: kartwheel/app/src/main/kotlin/us/handstand/kartwheel/KartWheel.kt
package us.handstand.kartwheel


import android.app.Application
import android.content.Context
import android.preference.PreferenceManager
import android.support.multidex.MultiDex
import com.crashlytics.android.Crashlytics
import io.fabric.sdk.android.Fabric
import us.handstand.kartwheel.controller.TicketController
import us.handstand.kartwheel.inject.DaggerInjector
import us.handstand.kartwheel.inject.Injector
import us.handstand.kartwheel.inject.provider.*
import us.handstand.kartwheel.layout.Font
import us.handstand.kartwheel.layout.GlideAppModule
import us.handstand.kartwheel.model.CompiledStatements
import us.handstand.kartwheel.model.Database
import us.handstand.kartwheel.model.Storage
import us.handstand.kartwheel.network.API
import us.handstand.kartwheel.notifications.PubNubManager
import us.handstand.kartwheel.util.Audio
import us.handstand.kartwheel.util.ThreadManager

open class KartWheel : Application() {

    override fun attachBaseContext(base: Context?) {
        super.attachBaseContext(base)
        if (BuildConfig.DEBUG) {
            MultiDex.install(this)
        }
    }

    override fun onCreate() {
        super.onCreate()
        Storage.initialize(PreferenceManager.getDefaultSharedPreferences(this))
        Database.initialize(this)
        CompiledStatements.initialize(Database.get())
        API.db = Database.get()
        Fabric.with(this, Crashlytics())
        Audio.initialize(this)
        injector = DaggerInjector.builder()
                .applicationProvider(ApplicationProvider(this))
                .controllerProvider(ControllerProvider())
                .cloudStorageProvider(CloudStorageProvider())
                .apiProvider(ApiProvider(BuildConfig.SERVER))
                .bottomSheetCallbackProvider(BottomSheetCallbackProvider())
                .build()

        injector.inject(API)
        injector.inject(Font)
    }

    companion object {
        lateinit var injector: Injector
        fun logout(submitOnExecutor: Boolean = true, lastTicketState: Long = TicketController.TOS) {
            if (submitOnExecutor) {
                ThreadManager.databaseExecutor.submit {
                    Storage.clear()
                    Database.clear(Database.get())
                    PubNubManager.tearDown()
                    Storage.lastTicketState = lastTicketState
                }
            } else {
                Storage.clear()
                Database.clear(Database.get())
                PubNubManager.tearDown()
                Storage.lastTicketState = lastTicketState
            }
        }
    }
}
